src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -582,6 +584,7 @@ public static ulong GetKey(ReadOnlySpan<byte> propertyName)
                 return ClassType.Enumerable;
             }
 
+            Debug.Assert(type.FullName! != null);
```

Comment:
Is this a valid assert, @steveharter, @layomia?

Reply:
Typo after FullName, extra `!` will be removed

Reply:
nit: There is an edge case where `type.FullName` can be null (resulting in a null ref here). I don't see such code be relevant to the serializer, but I also don't see any harm in guarding against null here.

Please change this from a Debug.Assert to an actual null check.

```C#
if (type.FullName != null)
{
  if (type.FullName.StartsWith...)
 
  // Also the other code that uses type.FullName
  ...
}
```

Here's a test:
```C#
public class ReadValueTests
{
   private static void MyMethod<T>(List<T> param)
   {

   }
   public static void TypeFullNameNullTest()
   {
        MethodInfo[] methods = typeof(ReadValueTests).GetMethods(BindingFlags.Static | BindingFlags.NonPublic);
        var parameters = methods[0].GetParameters();
        Console.WriteLine(parameters[0].ParameterType.FullName == null);

         // This currently throws null ref
        JsonSerializer.Deserialize("{}", parameters[0].ParameterType);
   }
}
```

Reply:
After null check now throwing: 
`System.ArgumentException : Cannot create an instance of System.Text.Json.JsonPropertyInfoNotNullable'4[System.Object,System.Collections.Generic.List'1[T],System.Collections.Generic.List'1[T],System.Collections.Generic.List'1[T]] because Type.ContainsGenericParameters is true.`
i guess it is expected, would assert that it throws `ArgumentException `